astropy/nddata/nddata.py

```diff
@@ -198,17 +213,23 @@ def __init__(
                 else:
                     mask = data.mask
 
-                # Just save the data for further processing, we could be given
-                # a masked Quantity or something else entirely. Better to check
-                # it first.
-                data = data.data
-
-            if isinstance(data, Quantity):
-                if unit is not None and unit != data.unit:
-                    log.info("overwriting Quantity's current unit with specified unit.")
-                else:
-                    unit = data.unit
-                data = data.value
+                # get the (possibly masked quantity) data from CCDData
+                # data = data.data
+
+            # if data inherits from ndarray, continue:
+            mro = data.__class__.__mro__
+            if np.ndarray in mro:
+                # if data is a MaskedQuantity, then the first base
+                # is the quantity, second is the numpy ndarray. The
+                # third would be a flat ndarray, but we don't want to
+                # go that deep, so stop getting bases:
+                while np.ndarray not in mro[:2]:
+                    if data.base is not None:
+                        data = data.base
+                    if issubclass(data.__class__, Quantity):
+                        unit = data.unit
+                        data = data.value
+                    mro = data.__class__.__mro__
```

Comment:
This block handles masked quantities and other Masked `np.ndarray` subclasses. Suppose you initialize `NDDataArray` like so
```python
import numpy as np
import astropy.units as u
from astropy.utils.masked import Masked

a = np.array([1, 2, 3, 4, 5])
mask = [True, False, False, True, True]
mqa = Masked(u.Quantity(a, u.m), mask=mask)
nd1 = NDDataArray(mqa)
```
then `NDDataArray` should get the values, mask, and unit from `mqa`. This turns out to be a bit involved, because `type(mqa)` is `astropy.utils.masked.core.MaskedQuantity`, and we have to use the `Masked` API to get the data values. To get the underlying quantity without a mask you can do `mqa.base`, which returns `[1, 2, 3, 4, 5] m`. To get the underlying data values in that quantity you can do `mqa.base.base` or `mqa.base.value`, which both return `array([1., 2., 3., 4., 5.])`. 

The highlighted block allows us to repeatedly go up the inheritance tree of `data` by taking `data.base` until the underlying `ndarray` and its associated masks and units are extracted.

Reply:
And since the comment in the highlighted block might not be clear, I'll elaborate a bit more here. Suppose you have an array with >1 dimensions: 
```python
>>> arr = np.arange(10).reshape((5, 2))
```
`arr` will have a `base` attribute, which is the raveled array:
```python
>>> arr.base
array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
```
In the first draft of the `while` loop above, I tried iteratively taking `data = data.base` until `data.base is None`, but the example in this comment shows that the `while` loop would ultimately ravel ND arrays (which isn't what I want).

Instead, the `while` loop will continue to reassign `data = data.base` until `data` is a Quantity. At that point the first two classes in the inheritance list `mro[:2]` are `[Quantity, np.ndarray]` and the loop will terminate after taking one last `data.base`. 

The loop may seem overly complicated, except that I wanted to be able to support subclasses whose `base` (or their `base`'s `base`, etc) is a (Masked)Quantity.

Reply:
This looks rather fragile to me. Maybe within `astropy`, it is OK to just rely on 
```
if isinstance(data, Quantity):
   unit = data.unit
   data = data.value
```
After that `value` is guaranteed to be either `ndarray` or `MaskedNDArray`.

This would not cover `Column`, though. Another option would be more simple,
```
unit = getattr(data, 'unit', None)
mask = getattr(data, 'mask', None)
data = np.asarray(data)
```
I think that is most consistent with duck-typing (except assumes `ndarray` -- could do `isinstance(data, np.ndarray)` before the last line.

Reply:
I'm not sure I follow, which portion of the PR would these examples replace?

> After that value is guaranteed to be either ndarray or MaskedNDArray.

Maybe I'm misunderstanding. Part of my intention in this PR is to split the values into `NDData.data`, the unit into `NDData.unit`, and the mask into `NDData.mask`. In your suggestion, if you take `data.value` and possibly get a `MaskedNDArray`, then the `data` attribute is itself a masked array. Wouldn't it be redundant and confusing to have both a `NDData.mask` attribute and a possibly different one in `NDData.data.mask`?

Reply:
I meant that the first step would get the unit and remove the `Quantity` aspect, then you could add a next step that got the `mask` and removed the `Masked` aspect.

But my second suggestion may be the easier one, as it would also cover `[Masked]Column`.

Reply:
@mhvk In the latest batch of commits (a7d3447 and later), I've replaced the highlighted block above with more explicit type checking for `data`, catching and handling each of the types we want to support here. For completeness, here's the full block that replaces the highlighted block above: 

https://github.com/astropy/astropy/blob/d7fd4036ecb3ed30dfcc8fcd0dbdfe486204d129/astropy/nddata/nddata.py#L193-L250

Reply:
Yes, that looks good!

Reply:
Thanks @mhvk! Might you have time to fully review this PR? I'm looking for volunteers 🙏🏻